### DISCUSSION TITLE
time diagnostics

### DISCUSSION THREAD

**Discussion by JamiePringle** [Parcels-code/Parcels#1293] [Q&A]
Is there anyway to have ocean parcels keep track of how much time it is doing in input, trajectory calculation, and output? I am trying to optimize run times, and am curious where it is spending the most time. 

Thanks,
Jamie

**Comment by erikvansebille** [comment]
Hi @JamiePringle, good question. @CKehl has developed a lot of code to do exactly this (for e.g. the preprint at https://arxiv.org/pdf/2105.00057.pdf). Most of that code sits in PR #1104, but that hasn't been merged yet. So the short answer is: not easy at the moment to find out where the time and memory burdens are in Parcels. 

Of course, you could run your own external profiler, like `cProfile`? Not as detailed and accurate as a built-in profiler, but for simple separation between input, execution and output it could be sufficient?

**Comment by JamiePringle** [comment]
Yes, cProfile and lineprofiler can be used. But wrapping the IO and particle advection routines in time accumulators makes it easier for the clueless user (:-) to understand what is going on. 

Basically, I was being lazy and trying not to re-invent the wheel....

Jamie
